server/frontend.go

```diff
@@ -3,68 +3,99 @@ package server
 import (
 	"log"
 	"os"
+	"sync/atomic"
+	"time"
 
 	"github.com/privacylab/talek/common"
 )
 
 // Frontend terminates client connections to the leader server.
+// It is the point of global serialization, and establishes sequence numbers.
 type Frontend struct {
 	// Private State
-	log          *log.Logger
-	name         string
-	serverConfig *Config
-	follower     common.FollowerInterface
-	isLeader     bool
+	log  *log.Logger
+	name string
+	*Config
 
-	//dataLayerRef *DataLayerRef
+	proposedSeqNo uint64 // Use atomic.AddUint64, atomic.LoadUint64
+	readChan      chan *readRequest
+
+	replicas []common.ReplicaInterface
+	dead     int
+}
+
+// readRequest is the grouped request and reply memory used for batching
+// incoming reads onto a single thread.
+type readRequest struct {
+	Args  *common.EncodedReadArgs
+	Reply *common.ReadReply
+	Done  chan bool
 }
 
 // NewFrontend creates a new Frontend for a provided configuration.
-func NewFrontend(name string, serverConfig *Config, follower common.FollowerInterface, isLeader bool) *Frontend {
+func NewFrontend(name string, config *Config, replicas []common.ReplicaInterface) *Frontend {
 	fe := &Frontend{}
 	fe.log = log.New(os.Stdout, "[Frontend:"+name+"] ", log.Ldate|log.Ltime|log.Lshortfile)
 	fe.name = name
-	fe.serverConfig = serverConfig
-	fe.follower = follower
-	fe.isLeader = isLeader
+	fe.Config = config
```

Comment:
Let's just make these variables thread-safe now right? We previously used atomic.Value to protect this, since we assume it can be changed in the future